python_modules/libraries/dagster-tableau/dagster_tableau/resources.py

```diff
@@ -0,0 +1,314 @@
+        pod_name: str,
+    ):
+        self.pod_name = pod_name
+        super().__init__(
+            connected_app_client_id=connected_app_client_id,
+            connected_app_secret_id=connected_app_secret_id,
+            connected_app_secret_value=connected_app_secret_value,
+            username=username,
+            site_name=site_name,
+        )
+
+    @property
+    def rest_api_base_url(self) -> str:
+        """REST API base URL for Tableau Cloud."""
+        return f"https://{self.pod_name}.online.tableau.com/api/{TABLEAU_REST_API_VERSION}"
+
+    @property
+    def metadata_api_base_url(self) -> str:
+        """Metadata API base URL for Tableau Cloud."""
+        return f"https://{self.pod_name}.online.tableau.com/api/metadata/graphql"
+
+
+@experimental
+class TableauServerClient(BaseTableauClient):
+    """Represents a client for Tableau Server and provides utilities
+    to interact with Tableau APIs.
+    """
+
+    def __init__(
+        self,
+        connected_app_client_id: str,
+        connected_app_secret_id: str,
+        connected_app_secret_value: str,
+        username: str,
+        site_name: str,
+        server_name: str,
+    ):
+        self.server_name = server_name
+        super().__init__(
+            connected_app_client_id=connected_app_client_id,
+            connected_app_secret_id=connected_app_secret_id,
+            connected_app_secret_value=connected_app_secret_value,
+            username=username,
+            site_name=site_name,
+        )
+
+    @property
+    def rest_api_base_url(self) -> str:
+        """REST API base URL for Tableau Server."""
+        return f"https://{self.server_name}/api/{TABLEAU_REST_API_VERSION}"
+
+    @property
+    def metadata_api_base_url(self) -> str:
+        """Metadata API base URL for Tableau Server."""
+        return f"https://{self.server_name}/api/metadata/graphql"
+
+
+@experimental
+class BaseTableauWorkspace(ConfigurableResource):
+    """Base class to represent a workspace in Tableau and provides utilities
+    to interact with Tableau APIs.
+    """
+
+    connected_app_client_id: str = Field(
+        ..., description="The client id of the connected app used to connect to Tableau Workspace."
+    )
+    connected_app_secret_id: str = Field(
+        ..., description="The secret id of the connected app used to connect to Tableau Workspace."
+    )
+    connected_app_secret_value: str = Field(
+        ...,
+        description="The secret value of the connected app used to connect to Tableau Workspace.",
+    )
+    username: str = Field(..., description="The username to authenticate to Tableau Workspace.")
+    site_name: str = Field(..., description="The name of the Tableau site to use.")
+
+    _client: Union[TableauCloudClient, TableauServerClient] = PrivateAttr(default=None)
+
+    def setup_for_execution(self, context: InitResourceContext) -> None:
+        self.build_client()
+
+    @abstractmethod
+    def build_client(self) -> None:
```

Comment:
I think it's worth deferring building the client until it's requested - you already do this in the context manager

Reply:
That makes sense - done in [d579257](https://github.com/dagster-io/dagster/pull/24233/commits/d579257a355ea9b4f3635be22ba9db7f59d31535)